File: sys/src/navigation/combined_imu_factor.cc
#include "combined_imu_factor.h"
#include "../base/rust.hpp"

namespace gtsam
{

    std::shared_ptr<CombinedImuFactor> default_combined_imu_factor()
    {
        return std::make_shared<CombinedImuFactor>();
    }

    std::shared_ptr<CombinedImuFactor> new_combined_imu_factor(
        Key pose_i, Key vel_i, Key pose_j, Key vel_j, Key bias_i, Key bias_j,
        const PreintegratedCombinedMeasurements & preintegratedMeasurements)
    {
        return std::make_shared<CombinedImuFactor>(
            pose_i, vel_i, pose_j, vel_j, bias_i, bias_j, preintegratedMeasurements);
    }

    std::shared_ptr<PreintegrationCombinedParams> new_preintegrated_combined_params_makesharedu()
    {
        boost::shared_ptr<gtsam::PreintegrationCombinedParams> params;
        params = PreintegrationCombinedParams::MakeSharedU();

        return to_std_ptr(params);
    }

    void set_accelerometer_covariance(
        std::shared_ptr<PreintegrationCombinedParams> &params,
        double sigma_a_sq)
    {
        params->setAccelerometerCovariance(gtsam::I_3x3 * sigma_a_sq);
    }

    void set_gyroscope_covariance(
        std::shared_ptr<PreintegrationCombinedParams> &params,
        double sigma_g_sq)
    {
        params->setGyroscopeCovariance(gtsam::I_3x3 * sigma_g_sq);
    }

    void bias_acc_covariance(
        std::shared_ptr<PreintegrationCombinedParams> &params,
        double sigma_wa_sq)
    {
        params->biasAccCovariance = sigma_wa_sq * gtsam::Matrix33::Identity(3, 3);
    }

    void bias_omega_covariance(
        std::shared_ptr<PreintegrationCombinedParams> &params,
        double sigma_wg_sq)
    {
        params->biasOmegaCovariance = sigma_wg_sq * gtsam::Matrix33::Identity(3, 3);
    }

    void set_integration_covariance(
        std::shared_ptr<PreintegrationCombinedParams> &params,
        double val)
    {
        params->setIntegrationCovariance(gtsam::I_3x3 * val);
    }

    void bias_acc_omega_int(
        std::shared_ptr<PreintegrationCombinedParams> &params,
        double val)
    {
        params->biasAccOmegaInt = val * gtsam::Matrix66::Identity(6, 6);
    }

    std::unique_ptr<PreintegratedCombinedMeasurements> default_preintegrated_combined_measurements()
    {
        return std::make_unique<PreintegratedCombinedMeasurements>();
    }

    std::unique_ptr<PreintegratedCombinedMeasurements> new_preintegrated_combined_measurements(
        const std::shared_ptr<PreintegrationCombinedParams> params,
        const imuBias::ConstantBias &bias)
    {
        // PreintegratedCombinedMeasurements * msmts = new PreintegratedCombinedMeasurements(to_boost_ptr(params));
        return std::make_unique<PreintegratedCombinedMeasurements>(to_boost_ptr(params), bias);
    }

    void integrateMeasurement(
        PreintegratedCombinedMeasurements &preintegrated_measurements,
        const Vector3 &measuredAcc,
        const Vector3 &measuredOmega,
        const double dt)
    {
        preintegrated_measurements.integrateMeasurement(measuredAcc, measuredOmega, dt);
    }

    std::unique_ptr<NavState> predict(
        const PreintegratedCombinedMeasurements &preintegrated_measurements,
        const NavState &state_i, const imuBias::ConstantBias & bias_i)
    {
        NavState state = preintegrated_measurements.predict(state_i, bias_i);
        return std::make_unique<NavState>(state);
    }

    void reset_integration_and_set_bias(
        PreintegratedCombinedMeasurements &preintegrated_measurements,
        const imuBias::ConstantBias &bias)
    {
            preintegrated_measurements.resetIntegrationAndSetBias(bias);
    }
}

File: src/navigation/combined_imu_factor.rs
use cxx::{UniquePtr, SharedPtr};
use crate::{base::vector::Vector3, imu::imu_bias::ConstantBias, inference::key::IntoKey};

use super::navstate::NavState;

pub struct CombinedImuFactor {
    pub inner: SharedPtr<::sys::CombinedImuFactor>,
}

impl Default for CombinedImuFactor {
    fn default() -> Self {
        Self {
            inner: ::sys::default_combined_imu_factor(),
        }
    }
}
impl CombinedImuFactor {
    pub fn new(
        pose_i: impl IntoKey, pose_j: impl IntoKey, vel_i: impl IntoKey, vel_j: impl IntoKey, 
        bias_i: impl IntoKey, bias_j: impl IntoKey, measurements: &PreintegratedCombinedMeasurements
    ) -> Self {
        Self {
            inner: ::sys::new_combined_imu_factor(
                pose_i.into_key(),
                pose_j.into_key(),
                vel_i.into_key(),
                vel_j.into_key(),
                bias_i.into_key(),
                bias_j.into_key(),
                & measurements.inner,
            ),
        }
    }
}

pub struct PreintegratedCombinedMeasurements {
    pub(super) inner: UniquePtr<::sys::PreintegratedCombinedMeasurements>,
}

impl Default for PreintegratedCombinedMeasurements {
    fn default() -> Self {
        Self {
            inner: ::sys::default_preintegrated_combined_measurements(),
        }
    }
}

impl PreintegratedCombinedMeasurements {
    pub fn new(params: PreintegrationCombinedParams, bias: &ConstantBias) -> Self {
        Self {
            inner: ::sys::new_preintegrated_combined_measurements(params.inner, &bias.inner),
        }
    }

    pub fn integrate_measurement(
        &mut self,
        measured_acc: &Vector3,
        measured_omega: &Vector3,
        dt: f64,
    ) {
        ::sys::integrateMeasurement(
            self.inner.pin_mut(),
            &measured_acc.inner,
            &measured_omega.inner,
            dt
        )
    }

    pub fn predict(
        &self,
        navstate: &NavState,
        bias: &ConstantBias,
    ) -> NavState {
        NavState {
            inner: ::sys::predict(&self.inner, &navstate.inner, &bias.inner),
        }
    }

    pub fn reset_integration_and_set_bias(
        &mut self,
        bias: &ConstantBias,
    ) {
        ::sys::reset_integration_and_set_bias(self.inner.pin_mut(), &bias.inner)
    }
}

pub struct PreintegrationCombinedParams {
    pub(super) inner: SharedPtr<::sys::PreintegrationCombinedParams>,
}

impl PreintegrationCombinedParams {
    pub fn makesharedu() -> Self {
        Self {
            inner: ::sys::new_preintegrated_combined_params_makesharedu(),
        }
    }
    pub fn set_params(&mut self, sigma_a_sq: f64, sigma_g_sq: f64, sigma_wa_sq: f64, sigma_wg_sq: f64, int_covar_val: f64, bias_acc_omega_int_val2: f64) {
        ::sys::set_accelerometer_covariance(&mut self.inner, sigma_a_sq);
        ::sys::set_gyroscope_covariance(&mut self.inner, sigma_g_sq);
        ::sys::bias_acc_covariance(&mut self.inner, sigma_wa_sq);
        ::sys::bias_omega_covariance(&mut self.inner, sigma_wg_sq);
        ::sys::set_integration_covariance(&mut self.inner, int_covar_val);
        ::sys::bias_acc_omega_int(&mut self.inner, bias_acc_omega_int_val2);
    }
}